hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -77,75 +73,66 @@ func createTypeConversion(
 	return nil, err
 }
 
-// assignPrimitiveTypeFromPrimitiveType will generate a direct assignment if both types have the
-// same underlying primitive type and are not optional
+// assignToOptionalType will generate a conversion where the destination is optional, if the
+// underlying type of the destination is compatible with the source.
 //
-// <destination> = <source>
+// <destination> = &<source>
 //
-func assignPrimitiveTypeFromPrimitiveType(
+func assignToOptionalType(
 	sourceEndpoint *StorageConversionEndpoint,
 	destinationEndpoint *StorageConversionEndpoint,
-	_ *StorageConversionContext) StorageTypeConversion {
-
-	// Require source to be non-optional
-	if _, srcOpt := AsOptionalType(sourceEndpoint.Type()); srcOpt {
-		return nil
-	}
-
-	// Require destination to be non-optional
-	if _, dstOpt := AsOptionalType(destinationEndpoint.Type()); dstOpt {
-		return nil
-	}
+	conversionContext *StorageConversionContext) StorageTypeConversion {
 
-	// Require source to be a primitive type
-	srcPrim, srcOk := AsPrimitiveType(sourceEndpoint.Type())
-	if !srcOk {
+	// Require destination to be optional
+	dstOpt, dstIsOpt := AsOptionalType(destinationEndpoint.Type())
+	if !dstIsOpt {
+		// Destination is not optional
 		return nil
 	}
 
-	// Require destination to be a primitive type
-	dstPrim, dstOk := AsPrimitiveType(destinationEndpoint.Type())
-	if !dstOk {
+	// Require a conversion between the wrapped type and our source
+	dstEp := destinationEndpoint.WithType(dstOpt.element)
+	conversion, _ := createTypeConversion(sourceEndpoint, dstEp, conversionContext)
+	if conversion == nil {
 		return nil
 	}
 
-	// Require both properties to have the same primitive type
-	if !srcPrim.Equals(dstPrim) {
-		return nil
-	}
+	//copyVar := destinationEndpoint.CreateLocal("", "Temp")
 
-	return func(reader dst.Expr, writer dst.Expr, generationContext *CodeGenerationContext) []dst.Stmt {
-		return []dst.Stmt{
-			astbuilder.SimpleAssignment(writer, token.ASSIGN, reader),
+	return func(reader dst.Expr, writer func(dst.Expr) []dst.Stmt, generationContext *CodeGenerationContext) []dst.Stmt {
+		// Create a writer that takes the address of the passed expression
+		// Note that we are dependent on any wrapping conversion to ensure no aliasing occurs
+		addrOfWriter := func(expr dst.Expr) []dst.Stmt {
+			return writer(astbuilder.AddrOf(expr))
```

Comment:
The writer is a func because it can be more complicated than just an assignment statement (e.g. for later conversions it can be a call). 

For example, in PR #429, I modify the writer here to check to see whether the `expr` passed in is a local variable - can't take the address of an arbitrary expression, so if `expr` is _not_ a local, we introduce one.